py/desispec/scripts/stdstars.py

```diff
@@ -688,23 +699,19 @@ def main(args, comm=None) :
         log.info('scaling {} mag {:.3f} to {:.3f} using scale {}'.format(ref_mag_name, model_magr, cur_refmag, scalefac))
         normflux[star] = model*scalefac
 
-    if comm is not None:
-        linear_coefficients = comm.reduce(linear_coefficients, op=MPI.SUM, root=0)
-        redshift = comm.reduce(redshift, op=MPI.SUM, root=0)
-        chi2dof = comm.reduce(chi2dof, op=MPI.SUM, root=0)
-        fitted_model_colors = comm.reduce(fitted_model_colors, op=MPI.SUM, root=0)
-        normflux = comm.reduce(normflux, op=MPI.SUM, root=0)
-
-    # Check at least one star was fit. The check is peformed on rank 0 and
-    # the result is bcast to other ranks so that all ranks exit together if
-    # the check fails.
-    atleastonestarfit = False
-    if rank == 0:
-        fitted_stars = np.where(chi2dof != 0)[0]
-        atleastonestarfit = fitted_stars.size > 0
-    if comm is not None:
-        atleastonestarfit = comm.bcast(atleastonestarfit, root=0)
-    if not atleastonestarfit:
+    if head_comm is not None and rank < nstars: # head_comm color is 1
+        linear_coefficients = head_comm.reduce(linear_coefficients, op=MPI.SUM, root=0)
+        redshift = head_comm.reduce(redshift, op=MPI.SUM, root=0)
+        chi2dof = head_comm.reduce(chi2dof, op=MPI.SUM, root=0)
+        fitted_model_colors = head_comm.reduce(fitted_model_colors, op=MPI.SUM, root=0)
+        normflux = head_comm.reduce(normflux, op=MPI.SUM, root=0)
+
+    if rank != 0: # All the work left belongs to rank 0
+        return
+
+    # Check at least one star was fit.
+    fitted_stars = np.where(chi2dof != 0)[0]
+    if fitted_stars.size == 0:
         log.error("No star has been fit.")
         sys.exit(12)
```

Comment:
I don't think we want to remove this, for the reason described in the comment. Also see here:

https://github.com/desihub/desispec/pull/1184#discussion_r594042298

Reply:
Yes, I think you are right. I added an early return before this check for all the ranks != 0, but I just realized that an early return does not gurantee that MPI won't hang on a statement outside of this function. I'll go ahead and add the broadcast back.

Reply:
Now I see that #1213 introduced a bug for the MPI stdstar fit after #1184 was merged. It looks like the code introduced in #1213 should have been under the `if rank == 0:` statement. Having ranks != 0 return before the write fixes a bug with the existing code. (Although, if rank 0 throws an exception during that final code block...)

Reply:
Yeah that is not ideal. We could consider using https://mpi4py.readthedocs.io/en/stable/mpi4py.run.html, which has an option to automatically call `MPI_Abort()` for any unhandled exceptions. This should also obviate the need to broadcast errors to all ranks.

Reply:
#1213 was my PR so if I broke it I'll take a look at fixing it.

Reply:
Documenting a separate conversation: the reason to not use `MPI_Abort()` is because that causes a hard exit of python for all ranks of all communicators without the ability to catch an exception.  This is problematic for workflow wrappers that might split a large communicator to process multiple exposure in parallel.  If there is a problem with one of those, then the wrapper should be able to catch the exception and move on without one exposure forcing the others to exit too.